### PR TITLE
Further transliteration of typing rules

### PR DESCRIPTION
**Pull request by jalglave**
@HadrienRenaud please could you have a look?
thanks in advance,
Jade

### CODE REVIEW

File: asllib/ASLTypingReference.mld [hunk 1]
```diff
@@ -722,124 +722,261 @@ of the following applies:
...
     {4 Example}
 
-  {3 TypingRule.SThen}
+  {3 TypingRule.SReturnOne}
 
     {4 Rule}
+    All of the following applies:
+    - One of the following applies:
+      - [s] is a return statement with some value and no return type;
+      - [s] is a return statement with no value and some return type;
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:810]: A generalization of my previous comment:
```suggestion
      - All of the following applies:
        - [s] is a return statement with some value;
        - the enclosing function does not have a return type;
      - All of the following applies:
        - [s] is a return statement with no value;
        - the enclosing function does have a return type;
```

File: asllib/ASLTypingReference.mld [hunk 2]
```diff
@@ -722,124 +722,261 @@ of the following applies:
...
+    - An error "Bad Return Statement" is raised.
 
     {4 Example}
 
-  {3 TypingRule.SCall}
+  {3 TypingRule.SReturnSome}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a return statement with some value [e] and some return type [t];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:819]:
```suggestion
    - [s] is a return statement with some value [e];
    - the enclosing function has a return type [t];
```

File: asllib/ASLTypingReference.mld [hunk 3]
```diff
@@ -722,124 +722,261 @@ of the following applies:
...
     {4 Rule}
+    All of the following applies:
+    - [s] is a for statement with index [id], direction [dir], two expressions
+      [e1] and [e2] and a statement [s'];
+    - [t1,e1'] is the result of annotating [e1] in [env];
+    - [t2,e2'] is the result of annotating [e2] in [env];
+    - [t1] has the structure of an integer type;
+    - [t2] has the structure of an integer type;
+    - [ty] is ;
+    - [s''] is the result of annotating [s'] in [env];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:902]:
```suggestion
    - [ty] is the integer type [integer {e1' .. e2'}];
    - [env1] is [env] with the identifier [id] locally declared as immutable of type [ty];
    - [s''] is the result of annotating [s'] in [env1];
```

Note that the type `ty` is actually more complicated as specified by rule ZSND (paragraph 6.9 of the LRM). We might want to specify this here.

File: asllib/ASLTypingReference.mld [hunk 4]
```diff
@@ -722,124 +722,261 @@ of the following applies:
...
 
     {4 Example}
 
-  {3 TypingRule.SThrowNone}
+  {3 TypingRule.SDeclSome}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a declaration with [ldk], [ldi] and an optional expression [e_opt];
+    - [e_opt] gives some expression [e];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:914]: I think this is equivalent and more explicit (we have a degree less of complexity)
```suggestion
    - [s] is a declaration with [ldk], [ldi] and an expression [e];
```

File: asllib/ASLTypingReference.mld [hunk 5]
```diff
@@ -722,124 +722,261 @@ of the following applies:
...
+    - [t_e,e'] is the result of annotating [e] in [env];
+    - [env', ldi'] is the result of declaring the local constants of [ldi] in [env];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:916]:
```suggestion
    - [env', ldi'] is the result of declaring the local identifiers of [ldi] in [env];
```

File: asllib/ASLTypingReference.mld [hunk 6]
```diff
@@ -722,124 +722,261 @@ of the following applies:
...
+    - [new_s] is a declaration with [ldk], [ldi'] and an expression [e'];
+    - [new_env] is [env].
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:918]:
```suggestion
    - [new_env] is [env'].
```

File: asllib/ASLTypingReference.mld [hunk 7]
```diff
@@ -722,124 +722,261 @@ of the following applies:
...
 
     {4 Example}
 
-  {3 TypingRule.SThrowSome}
+  {3 TypingRule.SDeclNone}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a declaration statement with [ldk], [ldi] and an optional expression [e_opt];
+    - [e_opt] gives no expression;
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:927]:
```suggestion
    - [s] is a declaration statement with [ldk], [ldi] and no initial expression;
```

File: asllib/ASLTypingReference.mld [hunk 8]
```diff
@@ -722,124 +722,261 @@ of the following applies:
...
+    - [env', s'] is the result of annotating uninitialised local declarations [ldi] in [env];
+    - [new_s] is [s'];
+    - [new_env] is [env].
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:930]:
```suggestion
    - [new_env] is [env'].
```

File: asllib/ASLTypingReference.mld [hunk 9]
```diff
@@ -722,124 +722,261 @@ of the following applies:
...
 
-  {3 TypingRule.SDeclSome}
+  {3 TypingRule.SThrowSome}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a throw statement with expression [e];
+    - [t_e,e'] is the result of annotating [e] in [env];
+    - [t_e] has the structure of an exception type;
+    - [new_s] is a throw statement with expression [e'] and type [t_e];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:951]: This is actually not syntax, so we might want to specify somewhere that we are adding some typing information in the AST.
Maybe `AST.mli` deserves an intro on that point?

File: asllib/ASLTypingReference.mld [hunk 10]
```diff
@@ -722,124 +722,261 @@ of the following applies:
...
     {4 Example}
 
   {3 TypingRule.SDebug}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a debug statement with expression [e];
+    - [_t_e, e'] is the result of annotating [e] in [env];
+    - [new_s] is a debug statement with expression [e'];
+    - [new_env] is [env].
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:976]: I do not think this should be in the spec. This is unspecified, and a very-asllib only syntax and code.

Probably once we fix the type-checking of the procedure `print`, all the `debug` statements could be converted in the ASL into `print` statements.

File: asllib/ASLTypingReference.mld [hunk 11]
```diff
@@ -981,54 +981,91 @@ one of the following applies:
...
 
     {4 Example}
 
   {3 TypingRule.SliceStar}
 
     {4 Rule}
+    All of the following applies:
+    - [slices] gives [(factor, pre_length)];
+    - [pre_offset] is [factor * pre_length];
+    - [offset, length] is the result of applying TypingRule.SliceLength to [(pre_offset, length)].
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1031]:
```suggestion
    - [offset, length] is the result of applying TypingRule.SliceLength to [(pre_offset, pre_length)].
```

File: asllib/ASLTypingReference.mld [hunk 12]
```diff
@@ -1041,30 +1078,92 @@ Formally, one of the following applies:
   {3 TypingRule.PGeq}
 
     {4 Rule}
+    All of the following applies:
+    - [p] is the pattern which matches anything greater than or equal to an expression [e];
+    - [t_e, e'] is the result of annotating [e] in [env]; 
+    - [e'] is pure;
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1084]: FIY, the LRM asks for a stronger check: that `e` (or `e'` similarly) is a compile-time constant.

File: asllib/ASLTypingReference.mld [hunk 13]
```diff
@@ -1041,30 +1078,92 @@ Formally, one of the following applies:
...
     {4 Rule}
+    All of the following applies:
+    - [p] is the pattern which matches anything within the range given by
+      expressions [e1] and [e2];
+    - [t_e1, e1'] is the result of annotating [e1] in [env];
+    - [t_e2, e2'] is the result of annotating [e2] in [env];
+    - [t] has the structure of an integer;
+    - [t_e1] has the structure of an integer;
+    - [t_e2] has the structure of an integer;
+    (*no need to check that e1' and e2' are pure here?*)
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1115]: Here we would need to check that `e1` and `e2` are compile-time constants..


I created HadrienRenaud/herdtools7#39 (and HadrienRenaud/herdtools7#40 which is related).

File: asllib/ASLTypingReference.mld [hunk 14]
```diff
@@ -1041,30 +1078,92 @@ Formally, one of the following applies:
...
     {4 Example}
 
   {3 TypingRule.PMask}
 
     {4 Rule}
+    All of the following applies:
+    - [p] is the pattern which matches a mask [m];
+    - [t] has the structure of a bitvector type;
+    - [n] is the length of mask [m];
+    - [t_m] is the bitvectore type of width [n];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1128]:
```suggestion
    - [t_m] is the bitvector type of width [n];
```

File: asllib/ASLTypingReference.mld [hunk 15]
```diff
@@ -1041,30 +1078,92 @@ Formally, one of the following applies:
...
+    - [t_m] type-satisfies [t];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1129]:
```suggestion
    - [t] type-satisfies [t_m];
```

File: asllib/ASLTypingReference.mld [hunk 16]
```diff
@@ -1041,30 +1078,92 @@ Formally, one of the following applies:
...
+
+    {4 Example}
+
+  {3 TypingRule.PTupleBadArity}
+
+    {4 Rule}
+    All of the following applies:
+    - [p] is the pattern which matches a tuple [li];
+    - [t_struct] is the type structure of [t];
+    - [t_struct] is a tuple [ts];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1140]:
```suggestion
    - [t] has the structure of a tuple type [ts];
```

File: asllib/ASLTypingReference.mld [hunk 17]
```diff
@@ -1041,30 +1078,92 @@ Formally, one of the following applies:
...
+    - an error "Bad Arity" is raised. 
 
     {4 Example}
 
   {3 TypingRule.PTuple}
 
     {4 Rule}
+    All of the following applies:
+    - [p] is the pattern which matches a tuple [li];
+    - [t_struct] is the type structure of [t];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1151]: FYI this is probably not what we exactly want: we would like `p_struct` to be the anonymous type corresponding to `t` in `env`

File: asllib/ASLTypingReference.mld [hunk 18]
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
...
+
+    {4 Info}
+    A local identifier declared with var, let or constant is in scope from the
+    point immediately after its declaration until the end of the immediately
+    enclosing block.
 
+    From that follows that we can discard the environment at the end of an
+    enclosing block.
+    
     {4 Example}
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1188]:
```suggestion
    {4 Example}
    [@asl{
    func main () => integer
    begin
      if UNKNOWN: boolean then
         let i = 3;
         print (i);
      end
      let i = "Some text";
      print (i);
      return 0;
    end
    }]
```

File: asllib/ASLTypingReference.mld [hunk 19]
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
...
-    <minimal example>
 
 {2
 Typing of Catchers
 }
-[annotate_catchers env catchers otherwise_opt s_m]
+  
+  {3 TypingRule.CatcherNone}
+  [annotate_catchers env return_type (name_opt, ty, stmt)] is [(name_opt, ty,
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1195]: It feels strange to me that `return_type` is not hidden in the local `env`. I might refacto that soon.

File: asllib/ASLTypingReference.mld [hunk 20]
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
...
+  {3 TypingRule.CatcherSome}
 
     {4 Rule}
-    <description>
+    All of the following applies:
+     - [ty] has the structure of an exception type;
+    - [name_opt] gives a name [name];
+    - [name] is not already declared in [env];
+    - [name] is annotated with [ty] in [env];
+    - [env'] is [env];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1216]:
```suggestion
    - [env'] is [env] with [name] locally declared as immutable of type [ty];
```

File: asllib/ASLTypingReference.mld [hunk 21]
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
...
+en name args eqs call_type] is [(name1, args, eqs2, ret_ty1)] or an error is
+raised and one of the following applies:
 
-  {3 TypingRule.FCall}
+  {3 TypingRule.FCallBadArity}
 
     {4 Rule}
-    <description>
+    All of the following applies:
+    - [caller_arg_types, arg1] is the result of annotating [args] in [env];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1234]:
```suggestion
    - [caller_arg_types, args1] is the result of annotating [args] in [env];
```

File: asllib/ASLTypingReference.mld [hunk 22]
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
...
+    - [extra_nargs, name1, callee_arg_types, ret_ty] is the result of trying to
+      find the name [name] with [caller_arg_types] in [env];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1236]: I find the following a bit more understandable, but I'm not sure it's that better:

```suggestion
    - [name] is bound in [env] to a function with argument types [callee_arg_types];
```

File: asllib/ASLTypingReference.mld [hunk 23]
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
...
+    - [eqs1] is the list made of both [eqs] and [extra_nargs];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1237]:
```suggestion
```

File: asllib/ASLTypingReference.mld [hunk 24]
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
...
+
+    {4 Example}
+
+  {3 TypingRule.FCallGetter}
+
+    {4 Rule}
+    All of the following applies:
+    - [caller_arg_types, arg1] is the result of annotating [args] in [env];
+    - [extra_nargs, name1, callee_arg_types, ret_ty] is the result of trying to
+      find the name [name] with [caller_arg_types] in [env];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1249]:
```suggestion
    - [name] is bound in [env] to a function with a unique name [name1] whose argument types [callee_arg_types] type-clash [caller_arg_types] and whose return type is [ret_ty];
```

File: asllib/ASLTypingReference.mld [hunk 25]
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
...
+    - [eqs1] is the list made of both [eqs] and [extra_nargs];
+    - [eqs2] is ;
+    - [call_type] is a either a function or a getter type;
+    - [ret_ty] is a type [ty];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1253]:
```suggestion
```

File: asllib/ASLTypingReference.mld [hunk 26]
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
...
+
+  {3 TypingRule.FCallGetter}
+
+    {4 Rule}
+    All of the following applies:
+    - [caller_arg_types, arg1] is the result of annotating [args] in [env];
+    - [extra_nargs, name1, callee_arg_types, ret_ty] is the result of trying to
+      find the name [name] with [caller_arg_types] in [env];
+    - [eqs1] is the list made of both [eqs] and [extra_nargs];
+    - [eqs2] is ;
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1251]:
```suggestion
    - [eqs2] is [eqs1] with added all the type equations from the type-clash of [caller_arg_types] and [callee_arg_types];
```

File: asllib/ASLTypingReference.mld [hunk 27]
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
...
 
 {3 TypingRule.Func}
 
   {4 Rule}
+  All of the following applies:
+  - [env1] is [env] modified to have an empty local environment;
+  - [env2] is ;
+  - [env3] is [env2] modified to add explicit parameters;
+  - [env4] is [env3] modified to resolve dependently typed identifiers in the arguments;
+  - [env5] is [env4] modified to resolve dependently typed identifiers in the result type;
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1301]: Are we still doing that? That shouldn't be the case.

File: asllib/ASLTypingReference.mld [hunk 28]
```diff
@@ -1073,49 +1172,136 @@ Typing of Blocks
...
-[name] <description> and one of the following applies:
+[annotate_func loc env f] annotates the function named [f] and returns [f],
+[new_body] and [name].
 
 {3 TypingRule.Func}
 
   {4 Rule}
+  All of the following applies:
+  - [env1] is [env] modified to have an empty local environment;
+  - [env2] is ;
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:1298]:
```suggestion
  - [env2] is [env1] with every formal argument declared as immutable with its type;
```

**Review by HadrienRenaud — COMMENTED** [review]
Hi @jalglave,

I will have to come back to some points on this review, as this points to a certain number of TODOs on my side.
There are certain comments that apply to other very similar rules.

Feel free to argue me out of those suggestions!

**Comment by jalglave** [comment]
thanks @HadrienRenaud I think I've addressed all your comments - please could you double-check, and if I've not missed anything, please could you merge?

**Comment by HadrienRenaud** [comment]
Hi @jalglave, I have marked as resolved the comments I think you've addressed. There are a few left, can you double check that you do not want them?
For example, this one looks like you've missed something: https://github.com/herd/herdtools7/pull/671/files#r1334979702

Otherwise, I'm happy to run tests, squash and merge when you want.

File: asllib/ASLTypingReference.mld (hunk 27 above)
Comment (HadrienRenaud) [reply asllib/ASLTypingReference.mld:1295]: Raised a new issue: https://github.com/HadrienRenaud/herdtools7/issues/41

**Pull request merged**